lib/active_record_view_model/update_data.rb

```diff
@@ -218,11 +280,25 @@ def parse(hash_data, valid_reference_keys)
             end
 
             if association_data.collection?
-              unless value.is_a?(Array)
-                raise ViewModel::DeserializationError.new("Could not parse non-array collection association")
-              end
+              associations[name] =
+                case value
+                when Array
+                  children = value.map { |child_hash| parse_association.(child_hash) }
+                  CollectionUpdate.new(CollectionUpdate::Type::Replace, children)
+
+                when Hash
+                  UpdateData.verify_schema!(Schemas::COLLECTION_UPDATE, value)
+                  functional_updates = value[ACTIONS_ATTRIBUTE].map do |action|
+                    type   = FunctionalUpdate::Type.parse!(action[TYPE_ATTRIBUTE])
+                    values = action[VALUES_ATTRIBUTE].map(&parse_association)
```

Comment:
It should be reference only IMHO. Which is how this implementation behaves when it comes to `build!`, but it's currently silent about it. Will add more input validation.